### PR TITLE
specs: use lazy lexicographic comparison instead of key_ordering

### PR DESCRIPTION
**Pull request by tgamblin**
We have been using the `@llnl.util.lang.key_ordering` decorator for specs and most of their components. This leverages the fact that in Python, tuple comparison is lexicographic. It allows you to implement a `_cmp_key` method on your class, and have `__eq__`, `__lt__`, etc. implemented automatically using that key. For example, you might use tuple keys to implement comparison, e.g.:

```python
class Widget:
    # author implements this
    def _cmp_key(self):
        return (
            self.a,
            self.b,
            (self.c, self.d),
            self.e
        )

    # operators are generated by @key_ordering
    def __eq__(self, other):
        return self._cmp_key() == other._cmp_key()

    def __lt__(self):
        return self._cmp_key() < other._cmp_key()

    # etc.
```

The issue there for simple comparators is that we have to build the tuples *and* we have to generate all the values in them up front. When implementing comparisons for large data structures, this can be costly.

This PR replaces `@key_ordering` with a new decorator, `@lazy_lexicographic_ordering`. Lazy lexicographic comparison maps the tuple comparison shown above to generator functions. Instead of comparing based on pre-constructed tuple keys, users of this decorator can compare using elements from a generator. So, you'd write:

```python
@lazy_lexicographic_ordering
class Widget:
    def _cmp_iter(self):
        yield a
        yield b
        def cd_fun():
            yield c
            yield d
        yield cd_fun
        yield e

    # operators are added by decorator (but are a bit more complex)
```

There are no tuples that have to be pre-constructed, and the generator does not have to complete. Instead of tuples, we simply make functions that lazily yield what would've been in the tuple. If a yielded value is a `callable`, the comparison functions will call it and recursively compare it. The comparator just walks the data structure like you'd expect it to.

The ``@lazy_lexicographic_ordering`` decorator handles the details of implementing comparison operators, and the ``Widget`` implementor only has to worry about writing ``_cmp_iter``, and making sure the elements in it are also comparable.

Using this PR shaves another 1.5 sec off the runtime of `spack buildcache list`, and it also speeds up Spec comparison by about 30%. The runtime improvement comes mostly from two things:

1. lazily stopping the comparison as soon as possible (e.g., many specs just have different names, which is the firs thing the lazy generators return)
2. avoiding the use of `hash()`  in `_cmp_iter()` (it was used in `_cmp_key()` before)

### CODE REVIEW

File: lib/spack/llnl/util/lang.py [hunk 1]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+#: sentinel for testing that iterators are done
+done = object()
+
+
+def tuplify(seq):
+    """Helper for lazy_lexicographic_ordering()."""
+    return tuple(tuplify(x) if callable(x) else x for x in seq())
+
+
+def doublewrap(f):
```
Comment (alalazo) [inline lib/spack/llnl/util/lang.py]: Minor, but using `decorator` instead of `f` may improve readability:
```suggestion
def doublewrap(decorator):
```

**Comment by tgamblin** [comment]
Seems there's a bug in here somewhere -- I might not get to fixing that until later today.

File: lib/spack/llnl/util/lang.py [hunk 2]
```diff
@@ -291,8 +452,9 @@ def __len__(self):
     def __delitem__(self, key):
         del self.dict[key]
 
-    def _cmp_key(self):
-        return tuple(sorted(self.values()))
+    def _cmp_iter(self):
+        for _, v in sorted(self.items()):
```
Comment (alalazo) [inline lib/spack/llnl/util/lang.py:474]: I assume the ordering is arbitrary for us, but want to note that this is different from the previous one (the ordering is done here on `(key, value)` pairs, while before it was just the values)

File: lib/spack/spack/test/spec_dag.py [hunk 3]
```diff
@@ -519,6 +519,11 @@ def test_normalize_mpileaks(self):
         assert spec == expected_flat
         assert spec.eq_dag(expected_flat)
 
+        print(spec.tree(color=True))
+        print(expected_normalized.tree(color=True))
+        print(spec == expected_normalized)
+        print(spec != expected_normalized)
```
Comment (alalazo) [inline lib/spack/spack/test/spec_dag.py]: Are these debug prints to be removed?

File: lib/spack/spack/test/spec_syntax.py [hunk 4]
```diff
@@ -828,4 +828,6 @@ def test_compare_abstract_specs(self):
 
         for a, b in itertools.product(specs, repeat=2):
             # Check that we can compare without raising an error
+            print("a:", str(a), type(a))
+            print("b:", str(b), type(b))
```
Comment (alalazo) [inline lib/spack/spack/test/spec_syntax.py]: Same question as above.

**Review by alalazo — DISMISSED** [review]
Minor comments from a first read of the PR. I'll test drive it asap.

File: lib/spack/llnl/util/lang.py [hunk 5]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+
+    Allows the decorator to be used with or without arguments, e.g.::
+
+        @decorator(with, arguments, and=kwargs)
 
-       If a class already has ``__eq__``, ``__ne__``, ``__lt__``, ``__le__``,
-       ``__gt__``, or ``__ge__`` defined, this decorator will overwrite them.
+    or::
+
+        @decorator
```
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py:244]:
```suggestion
        @decorator  # Calls the function with zero arguments.
```

File: lib/spack/llnl/util/lang.py (hunk 1 above)
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py]:
```suggestion
def coerce_decorator_arguments(f):
```
along with @alalazo's suggestion!

File: lib/spack/llnl/util/lang.py [hunk 6]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+
+    See https://stackoverflow.com/questions/653368.
```
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py]: I would usually make this a comment just below the docstring, as it seems like an implementation detail.

File: lib/spack/llnl/util/lang.py [hunk 7]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+            def __lt__(self):
+                return self._cmp_key() < other._cmp_key()
+
+            # etc.
+
+    Python would compare ``Widgets`` lexicographically based on their
+    tuples. The issue there for simple comparators is that we have to
+    bulid the tuples *and* we have to generate all the values in them up
+    front. When implementing comparisons for large data structures, this
+    can be costly.
```
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py:350]: Should we be using this whenever possible, then? Is there a use case where this doesn't apply?

File: lib/spack/llnl/util/lang.py [hunk 8]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+    to complete. Instead of tuples, we simply make functions that lazily
+    yield what would've been in the tuple. The
+    ``@lazy_lexicographic_ordering`` decorator handles the details of
+    implementing comparison operators, and the ``Widget`` implementor
+    only has to worry about writing ``_cmp_iter``, and making sure the
+    elements in it are also comparable.
+
+    Note: if a class already has ``__eq__``, ``__ne__``, ``__lt__``,
+    ``__le__``, ``__gt__``, ``__ge__``, or ``__hash__`` defined, this
+    decorator will overwrite them.
```
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py]: Could we make it raise a `TypeError` instead, unless a keyword argument is provided to the decorator? I'm concerned silently overwriting methods by default could be difficult to debug at some point.

File: lib/spack/llnl/util/lang.py [hunk 9]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+
+    add_func_to_class(
+        "__eq__", lambda s, o:
+        (s is o) or (o is not None and eq(s._cmp_iter, o._cmp_iter)))
+    add_func_to_class(
+        "__ne__", lambda s, o:
+        (s is not o) and (o is None or ne(s._cmp_iter, o._cmp_iter)))
+
+    add_func_to_class(
+        "__lt__", lambda s, o:
```
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py]: If we break these `lambda`s out into `def`s and use `@memoized`, does that affect performance? Could it affect performance in the future?

File: lib/spack/spack/spec.py [hunk 10]
```diff
@@ -4225,6 +4231,16 @@ def target(self):
         # to give to the attribute the appropriate comparison semantic
         return self.architecture.target.microarchitecture
 
+    def __hash__(self):
+        """Memoized version of lazy_lexicographic_ordering hash."""
```
Comment (cosmicexplorer) [inline lib/spack/spack/spec.py]: Is there a reason this isn't implemented in the `lazy_lexicographic_ordering` decorator? It seems useful for a lot more than the Spec class.

File: lib/spack/llnl/util/lang.py [hunk 11]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
-    setter('__ne__',
-           lambda s, o:
-           (s is not o) and (o is None or s._cmp_key() != o._cmp_key()))
-    setter('__gt__',
-           lambda s, o: o is None or s._cmp_key() > o._cmp_key())
-    setter('__ge__',
-           lambda s, o: o is None or s._cmp_key() >= o._cmp_key())
-
-    setter('__hash__', lambda self: hash(self._cmp_key()))
+    if not has_method(cls, "_cmp_iter"):
```
Comment (cosmicexplorer) [inline lib/spack/llnl/util/lang.py:395]: I might add a TODO here to try to check this with a mypy protocol before runtime (https://mypy.readthedocs.io/en/stable/protocols.html).

File: lib/spack/llnl/util/lang.py (hunk 8 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py]: It's hard to know this b/c every object has these methods:

```python
>>> o = object()
>>> hasattr(o, "__lt__")
True
```

I haven't seen other decorators that go to great lengths to avoid this -- if you use the decorator, it's pretty explicit how comparison should work.

File: lib/spack/llnl/util/lang.py (hunk 9 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py]: we don't want these to be memoized, as they could be used on objects that change (like specs).  `Spec` specifically memoizes the hash (because we already handle invalidation for that) but I don't think it should be the default.

File: lib/spack/spack/spec.py (hunk 10 above)
Comment (tgamblin) [reply lib/spack/spack/spec.py]: It is implemented there.  This one is an override.  Note that Spec is defined with:

```python
@lazy_lexicographic_ordering(set_hash=False)
class Spec(...):
...
```

File: lib/spack/llnl/util/lang.py [hunk 12]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+            )
+
+        return False  # if equal, return False
+
+    # other comparison operators are implemented in terms of eq and lt
+    def ne(lseq, rseq):
+        return not eq(lseq, rseq)
+
+    def gt(lseq, rseq):
+        return lt(lseq, rseq, operator.gt, right_shorter)
```
Comment (becker33) [inline lib/spack/llnl/util/lang.py]:
```suggestion
        return lt(rseq, lseq)
```
Then we can remove op and diff as arguments to `lt`.

File: lib/spack/llnl/util/lang.py [hunk 13]
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
...
+        setattr(cls, name, func)
+
+    add_func_to_class(
+        "__eq__", lambda s, o:
+        (s is o) or (o is not None and eq(s._cmp_iter, o._cmp_iter)))
+    add_func_to_class(
+        "__ne__", lambda s, o:
+        (s is not o) and (o is None or ne(s._cmp_iter, o._cmp_iter)))
+
+    add_func_to_class(
```
Comment (becker33) [inline lib/spack/llnl/util/lang.py]: Add a note that all objects are greater than `None` (using python2 semantics in python3).

File: lib/spack/spack/architecture.py [hunk 14]
```diff
@@ -329,23 +329,27 @@ def __repr__(self):
...
-                self.default_os,
-                self.front_os,
-                self.back_os,
-                t_keys,
-                o_keys)
-
-
-@key_ordering
+    def _cmp_iter(self):
+        yield self.name
```
Comment (becker33) [inline lib/spack/spack/architecture.py:333]: All yielded values need to `or` to something sane for python3 comparisons.

```suggestion
        yield self.name or ''
```
etc for all compared values.

File: lib/spack/spack/spec.py [hunk 15]
```diff
@@ -686,10 +689,10 @@ def add_type(self, type):
             self.deptypes + dp.canonical_deptype(type)
         )
 
-    def _cmp_key(self):
-        return (self.parent.name if self.parent else None,
-                self.spec.name if self.spec else None,
-                self.deptypes)
+    def _cmp_iter(self):
+        yield self.parent.name if self.parent else None
```
Comment (becker33) [inline lib/spack/spack/spec.py:693]: Can't compare `None` to `str` in python3

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/llnl/util/lang.py (hunk 5 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py:244]: fixed

File: lib/spack/llnl/util/lang.py (hunk 6 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py]: fixed

File: lib/spack/llnl/util/lang.py (hunk 12 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py]: fixed

File: lib/spack/llnl/util/lang.py (hunk 13 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py]: I mentioned this in the docstring.

File: lib/spack/spack/architecture.py (hunk 14 above)
Comment (tgamblin) [reply lib/spack/spack/architecture.py:333]: I worked the None comparison semantics into lazy lexicographic ordering, so this is actually no longer necessary.  I think this makes our comparisons way easier to write.

**Comment by tgamblin** [comment]
@alalazo @becker33 @eugeneswalker: this is ready for another look.  With the latest E4S build cache (which has 37k specs), it significantly improves performance for `spack buildcache list`:

`develop`:

```console
       42.68 real        35.00 user         0.75 sys
```

`lazy-lexicographic-spec-comparison`:

```console
       24.71 real        23.80 user         0.46 sys
```

So that shaves > 40% of the time off.

The key is really two things.  First, `__lt__` is faster b/c of the laziness -- specs that are actually less than others bail as early as possible.  For `__eq__`, or rather for specs that are equal, things are trickier, because we do not want `__eq__` to have to traverse an entire spec.  I'm leveraging the dag hash like this:

```python
    def __hash__(self):
        # If the spec is concrete, we leverage the DAG hash and just use
        # a 64-bit prefix of it. The DAG hash has the advantage that it's
        # computed once per concrete spec, and it's saved -- so if we
        # read concrete specs we don't need to recompute the whole hash.
        # This is good for large, unchanging specs.
        if self.concrete:
            if not self._dunder_hash:
                self._dunder_hash = self.dag_hash_bit_prefix(64)
            return self._dunder_hash

        # This is the normal hash for lazy_lexicographic_ordering. It's
        # slow for large specs because it traverses the whole spec graph,
        # so we hope it only runs on abstract specs, which are small.
        return hash(lang.tuplify(self._cmp_iter))
```

So basically, I think this gets us the best of both worlds -- lazy comparison for abstract specs that are small, and leveraging precomputed hashes for specs that are large (which are pretty much all the concrete ones).

**Review by eugeneswalker — DISMISSED** [review]
This looks great! Thank you.

File: lib/spack/spack/test/spec_syntax.py [hunk 16]
```diff
@@ -352,10 +352,10 @@ def test_multiple_specs_with_hash(self, database):
     def test_ambiguous_hash(self, mutable_database):
         x1 = Spec('a')
         x1.concretize()
-        x1._hash = 'xy'
+        x1._hash = 'xyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy'
```
Comment (becker33) [inline lib/spack/spack/test/spec_syntax.py:355]: Why is this necessary?

File: lib/spack/llnl/util/lang.py [hunk 17]
```diff
@@ -227,48 +232,224 @@ def list_modules(directory, **kwargs):
...
+    def ge(self, other):
+        if self is other:
+            return True
+        return (other is None) or not lazy_lt(self._cmp_iter, other._cmp_iter)
+
+    def h(self):
+        return hash(tuplify(self._cmp_iter))
+
+    # lambdas below handle special cases with None, then delegate to the
+    # recursive functions that take generators above.
```
Comment (becker33) [inline lib/spack/llnl/util/lang.py]: This comment is no longer accurate

**Review by becker33 — CHANGES_REQUESTED** [review]
Only request is removing a vestigial comment. Otherwise LGTM

File: lib/spack/spack/test/spec_syntax.py (hunk 16 above)
Comment (tgamblin) [reply lib/spack/spack/test/spec_syntax.py:355]: The fact that `Spec.__hash__` now uses [`self.dag_hash_bit_prefix(64)`](https://github.com/spack/spack/pull/21618/files#diff-9e28cdd6f9cd66e61101de6e9d22b1c33561959e4305c63147bbfbf7b22da936R4342-R4345) means that there need to be at least 64 bits of hash data to get the prefix.  That's 13 base32 characters, so I just went ahead and made these hashes full length.  Every real hash will be 32 characters long (160 bits for SHA1) so I made them proper stand-ins.

File: lib/spack/llnl/util/lang.py (hunk 17 above)
Comment (tgamblin) [reply lib/spack/llnl/util/lang.py]: fixed

**Comment by tgamblin** [comment]
@becker33:comments addressed

**Review by becker33 — APPROVED** [review]

**Comment by tgamblin** [comment]
@alalazo FYI

**Comment by alalazo** [comment]
Thanks I'll update #21683
